pocs/utils/google/storage.py

```diff
@@ -97,3 +99,247 @@ def upload(self, local_path, remote_path=None):
+
+        Args:
+            remote_path (str|`google.cloud.storage.blob.Blob`): Blob or path to remote blob.
+                If just the blob name is given then file will be downloaded.
+            save_dir (str, optional): Path for saved file, defaults to current directory.
+            force (bool, optional): Force a download even if file exists locally, default False.
+            callback (callable, optional): A callable object that gets called at end of
+                function.
+
+        Returns:
+            str: Path to local (uncompressed) FITS file
+        """
+        # Get blob object if just a path
+        if isinstance(remote_path, str):
+            remote_path = self.get_file_blob(remote_path)
+
+        output_path = os.path.join(
+            save_dir,
+            remote_path.name.replace('/', '_')
+        )
+
+        # Make dir if needed
+        os.makedirs(os.path.dirname(output_path), exist_ok=True)
+
+        # If we don't  have file (or force=True) then download directly to
+        # (compressed) FITS file.
+        if not os.path.exists(output_path) or force:
+            with open(output_path, 'wb') as f:
+                remote_path.download_to_file(f)
+
+        # Wait for download
+        timeout = 10
+        while os.path.exists(output_path) is False:
+            timeout -= 1
+            if timeout < 0:
+                return None
+            time.sleep(1)
+
+        # User supplied callback
+        if callback is not None:
+            try:
+                callback()
+            except TypeError:
+                warn('callback must be callable')
+
+        return output_path
+
+    def lookup_fits_header(self, remote_blob, parse_line=None):
+        """Read the FITS header from storage.
+
+        FITS Header Units are stored in blocks of 2880 bytes consisting of 36 lines
+        that are 80 bytes long each. The Header Unit always ends with the single
+        word 'END' on a line (not necessarily line 36).
+
+        Here the header is streamed from Storage until the 'END' is found, with
+        each line given minimal parsing.
+
+        See https://fits.gsfc.nasa.gov/fits_primer.html for overview of FITS format.
+
+        Args:
+            remote_blob (google.cloud.storage.blob.Blob): Blob object
+                corresponding to stored FITS file.
+            parse_line (callable, optional): A callable function which will be
+                passed an 80 character string corresponding to each line in the header.
+
+        Returns:
+            dict: FITS header as a dictonary.
+        """
+        i = 1
+        if remote_blob.name.endswith('.fz'):
+            i = 2  # We skip the compression header info
+
+        headers = dict()
+
+        streaming = True
+        while streaming:
+            # Get a header card
+            start_byte = 2880 * (i - 1)
+            end_byte = (2880 * i) - 1
+            b_string = remote_blob.download_as_string(start=start_byte, end=end_byte)
```

Comment:
Unless you know that the FITS header is usually just one block, I'd be inclined to simply download the first N blocks of size 2880 bytes (e.g. 50 blocks), then operate on each of those blocks (or really treat it as a longer sequence of 80 byte lines. This will very likely be faster than downloading two 2880 blocks.

Or test to see if I'm right.

Reply:
We know that the first block is the compression header information (for `.fz` packed files) so we are skipping that block and starting at block 2. After that we don't know how many blocks it will be.

I'll add some comments so it's clear what is trying to be done then we can figure out if that is what actually should be done. :)